proxy/src/keystore.rs

```diff
@@ -0,0 +1,211 @@
+//! Storage of secret keys.
+
+use std::convert::Infallible;
+use std::fmt;
+
+use secstr::{SecStr, SecUtf8};
+
+use librad::keys;
+use librad::paths;
+use radicle_keystore::{
+    crypto::{Pwhash, SecretBoxError},
+    file, FileStorage, Keystore, SecretKeyExt,
+};
+use radicle_registry_client::{ed25519, CryptoError, CryptoPair};
```

Comment:
As we start to depend on `ed25519` beyond full-filling registry related functionality we should move to explicitly depend on it down the road. Maybe worth capturing in a TODO or a follow-up issue. Alternatively the keystore crate exposes it.

Reply:
To expand on the comment above, it seems a common utility we need and we should have a unified way to handle keys, which in turn will benefit auditability and correctness. I know there is a strong resistance to blur the boundaries between the protocols, but this seems like a @radicle-dev/engineering concern.

Reply:
Maybe I was to quick with my ideas here, looks like we only use it for the registry key. Why do we use different keys? Thought we could tap into `ed25519` for both of them.

Reply:
I would be surprised if registry would also use `libsodium`. I agree, however, that it is a constant battle to not depend on `n > 1` implementations of the same signature algorithm, because crates tend to not abstract over it (and instead depend on a specific implementation).

The solution I would propose is that both `librad` and `radicle-registry-client`, instead of passing around secret key material, standardise on a trait like this:

```rust
pub trait Signer {
    type PublicKey;
    type Signature;

    fn public_key(&self) -> Self::PublicKey;
    fn sign(&self, data: &[u8]) -> Self::Signature;
}
```

Implementations of this trait could, for example, delegate signing to an `ssh-agent`, which could greatly improve usability.

Note that the dependency of `radicle-keystore` on `libsodium` is unaffected by this, and not in itself a problem -- the keystore doesn't perform signing, it uses key derivation and "sealing" acc. to the NaCl bestest practices.

Reply:
I second the proposal to use a `Signer` trait to do the signing. This should be straightforward to implement on the registry client side.

Reply:
Great summary and I think the proposal is sound. Can we capture the follow-ups?

Reply:
You want to do a follow-up or I make these changes here?

Reply:
From my understanding this would require follow-ups in both coco and registry, followed by integration of these follow-ups in the app. Happy to look into it.

Reply:
I have some follow-up questions on the design and how we want to tackle this consistently between the code bases. If we had a place to talk about this, @xla. I don’t care where exactly.